### PR TITLE
[deckhouse-controller] Move `tools/change-registry.sh` to `deckhouse-controller`

### LINKED ISSUE #4326: change-registry.sh will not work at version 1.45+ because of init containter pull change from tag to sha256
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Version

v1.45.0

### Expected Behavior

change-registry.sh changes registry correctly

### Actual Behavior

change-registry.sh will be broken after v1.45.0 because of change init-container image format from
`image: ${REGISTRY_ADDRESS}${REGISTRY_PATH}:${DECKHOUSE_INIT_EXTERNAL_MODULES_CONTAINER_TAG}`
to 
`image: ${REGISTRY_ADDRESS}${REGISTRY_PATH}@${DECKHOUSE_INIT_EXTERNAL_MODULES_CONTAINER_SHA256}`

### Steps To Reproduce

_No response_

### Additional Information

_No response_

### Logs

_No response_

### LINKED ISSUE #4328: change-registry.sh don't add imagePullSecrets in case of switching from CE to FE
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Version

v1.44.4

### Expected Behavior

change-registry.sh add imagePullSecrets in case of switching from CE to FE

### Actual Behavior

Have constantly ImagePullBackoff of Deckhouse pod

### Steps To Reproduce

_No response_

### Additional Information

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by alex123012**
## Description

Move `tools/change-registry.sh` to `deckhouse-controller`.

## Why do we need it, and what problem does it solve?

Fix #4328
Fix #4326

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: Move `tools/change-registry.sh` to `deckhouse-controller`.
impact_level: default
```

### CODE REVIEW

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 1]
```diff
@@ -339,24 +339,21 @@ proxy:
 
 To switch the Deckhouse cluster to using a third-party registry, follow these steps:
 
-* Update the `image` field in the `d8-system/deckhouse` deployment to contain the address of the Deckhouse image in the third-party-registry;
-* Download script to the master node and run it with parameters for a new registry.
+* Run `deckhouse-controller helper change-registry` inside `Deckhouse` pod with the new registry settings.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:342]:
```suggestion
* Run `deckhouse-controller helper change-registry` inside the `Deckhouse` Pod with the new registry settings.
```

File: docs/documentation/pages/DECKHOUSE-FAQ.md [hunk 2]
```diff
@@ -339,24 +339,21 @@ proxy:
...
   ```shell
-  curl -fsSL -o change-registry.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/change-registry.sh
-  chmod 700 change-registry.sh
-  ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+  kubectl exec -ti -n d8-system deploy/deckhouse -- deckhouse-controller helper change-registry --user my-user --password my-password registry.example.com/deckhouse
   ```
-  
-  * If the registry uses a self-signed certificate, put the root CA certificate that validates the registry's HTTPS certificate to file `ca.crt` near the script and add the `--ca-file ca.crt` option to the script.
+
+  * If the registry uses a self-signed certificate, put the root CA certificate that validates the registry's HTTPS certificate to file `ca.crt` in the deckhouse pod and add the `--ca-file ca.crt` option to the script.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ.md:349]:
```suggestion
  * If the registry uses a self-signed certificate, put the root CA certificate that validates the registry's HTTPS certificate to file `ca.crt` in the `deckhouse` Pod and add the `--ca-file ca.crt` option to the script.
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 3]
```diff
@@ -343,24 +343,21 @@ proxy:
 
 Для переключения кластера Deckhouse на использование стороннего registry выполните следующие действия:
 
-* Измените поле `image` в Deployment `d8-system/deckhouse` на адрес образа Deckhouse в новом registry;
-* Скачайте скрипт на мастер-узел и запустите его с параметрами нового registry.
+* Запустите `deckhouse-controller helper change-registry` из пода `Deckhouse` с параметрами нового registry.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:346]:
```suggestion
* Выполните команду `deckhouse-controller helper change-registry` из Пода `Deckhouse` с параметрами нового registry.
```

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 4]
```diff
@@ -343,24 +343,21 @@ proxy:
...
 
   ```shell
-  curl -fsSL -o change-registry.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/change-registry.sh
-  chmod 700 change-registry.sh
-  ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+  kubectl exec -ti -n d8-system deploy/deckhouse -- deckhouse-controller helper change-registry --user my-user --password my-password registry.example.com/deckhouse
   ```
 
-  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию `--ca-file ca.crt`.
+  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` в поде deckhouse и добавьте к вызову опцию `--ca-file ca.crt`.
```
Comment (z9r5) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:353]:
```suggestion
  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` в Поде `deckhouse` и добавьте к вызову опцию `--ca-file ca.crt`.
```

**Review by z9r5 — DISMISSED** [review]

File: deckhouse-controller/pkg/helpers/register.go [hunk 5]
```diff
@@ -35,6 +36,22 @@ func DefineHelperCommands(kpApp *kingpin.Application) {
 		})
 	}
 
+	{
+		changeRegistryCommand := helpersCommand.Command("change-registry", "Change registry for deckhouse images.")
+		newRegistry := changeRegistryCommand.Arg("new-registry", "Registry that will be used for deckhouse images (example: registry.example.com/deckhouse).").Required().String()
```
Comment (yalosev) [inline deckhouse-controller/pkg/helpers/register.go:41]: Could we use actual registry for example?
Also it's not obvious - do we need to pass the scheme or not
```suggestion
		newRegistry := changeRegistryCommand.Arg("new-registry", "Registry that will be used for deckhouse images (example: https://registry.deckhouse.io/deckhouse/ce).").Required().String()
```

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+// in the PR https://github.com/google/go-containerregistry/pull/1709
+// some methods became public and we can use them except of copying the source code.
+// (waiting for new "go-containerregistry" version)
+
+const (
+	d8SystemNS = "d8-system"
+)
+
+func ChangeRegistry(newRegistry, username, password, caFile, newDeckhouseImageTag string, insecure bool) error {
+	ctx := context.TODO()
```
Comment (yalosev) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:54]:
```suggestion
	ctx := context.Background()
```

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+func updateDeployment(ctx context.Context, kubeCl kclient.KubeClient, deploy *appsv1.Deployment) error {
+	deployClient := kubeCl.AppsV1().Deployments(deploy.Namespace)
+	updateOpts := metav1.UpdateOptions{FieldValidation: metav1.FieldValidationStrict}
+	if _, err := deployClient.Update(ctx, deploy, updateOpts); err != nil {
+		return err
+	}
+	return nil
+}
+
+func updateImageRepoForContainers(containers []v1.Container, newRepository string, nameOpts []name.Option, remoteOpts []remote.Option, newTagsForContainer map[string]string /* map[<container name>]>newTag> */) ([]v1.Container, error) {
```
Comment (yalosev) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:259]:
```suggestion
func updateImageRepoForContainers(containers []v1.Container, newRepository string, nameOpts []name.Option, remoteOpts []remote.Option, newTagsForContainer map[string]string /* map[<container name>]<newTag> */) ([]v1.Container, error) {
```

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+		if toChallenge(resp.Header.Get(wwwAuthHeader)) == bearer {
+			return nil
+		}
+		return fmt.Errorf("can't use bearer token auth with registry %s", reg.Name())
+	}
+
+	return errors.New(strings.Join(errs, "; "))
+}
+
+func toChallenge(s string) string {
```
Comment (yalosev) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:369]: Didn't get this. Do we really need this function?

File: docs/documentation/pages/DECKHOUSE-FAQ_RU.md [hunk 9]
```diff
@@ -343,24 +343,21 @@ proxy:
...
 
   ```shell
-  curl -fsSL -o change-registry.sh https://raw.githubusercontent.com/deckhouse/deckhouse/main/tools/change-registry.sh
-  chmod 700 change-registry.sh
-  ./change-registry.sh --registry-url https://my-new-registry/deckhouse --user my-user --password my-password
+  kubectl exec -ti -n d8-system deploy/deckhouse -- deckhouse-controller helper change-registry --user my-user --password my-password registry.example.com/deckhouse
   ```
 
-  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` возле скрипта и добавьте к вызову опцию `--ca-file ca.crt`.
+  * Если registry использует самоподписные сертификаты, то положите корневой сертификат соответствующего сертификата registry в файл `ca.crt` в Поде `deckhouse` и добавьте к вызову опцию `--ca-file ca.crt`.
```
Comment (yalosev) [inline docs/documentation/pages/DECKHOUSE-FAQ_RU.md:353]: Any way to pass CA from stdin? Or it's overcomplicated to realize and use?

**Review by z9r5 — APPROVED** [review]

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 10]
```diff
@@ -0,0 +1,367 @@
+// Copyright 2023 Flant JSC
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:1]: A unit test is required.

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+	keyBlock, _ := pem.Decode(caBytes)
+	if keyBlock == nil {
+		return "", errors.New("can't read CA file content as pem file")
+	}
+
+	// Check that file content is a certificate
+	if _, err := x509.ParseCertificate(keyBlock.Bytes); err != nil {
+		return "", err
+	}
+	return strings.Trim(string(caBytes), "\n "), nil
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:221]: strings.strings.TrimSpace(string(caBytes)), nil

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+		if err != nil {
+			errs = append(errs, err.Error())
+			// Potentially retry with http.
+			continue
+		}
+		defer func() {
+			// By draining the body, make sure to reuse the connection made by
+			// the ping for the following access to the registry
+			defer resp.Body.Close()
+			io.Copy(io.Discard, resp.Body)
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:346]: this is not required, `defer resp.Body.Close()` is enough

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+		}()
+
+		if resp.StatusCode != http.StatusUnauthorized {
+			return transport.CheckError(resp, http.StatusUnauthorized)
+		}
+
+		if challenges := authchallenge.ResponseChallenges(resp); len(challenges) != 0 {
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:353]: Let's avoid unnecessary docker distribution dependency. It should be enough to just copy the `ResponseChallenges` method.

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+			// If we hit more than one, I'm not even sure what to do.
+			wac := challenges[0]
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:355]: Potential panic.

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+
+	// This first attempts to use "https" for every request, falling back to http
+	// if the registry matches our localhost heuristic or if it is intentionally
+	// set to insecure via name.NewInsecureRegistry.
+	schemes := []string{"https"}
+	if reg.Scheme() == "http" {
+		schemes = append(schemes, "http")
+	}
+
+	var errs []string
```
Comment (nabokihms) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:329]: We are using hashicorp multi error for this.

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go (hunk 14 above)
Comment (yalosev) [reply deckhouse-controller/pkg/helpers/change_registry/change_registry.go:355]:
```
; len(challenges) != 0 
```
two lines above

**Pull request merged**